### DISCUSSION TITLE
log_system_metrics not working

### DISCUSSION THREAD

**Discussion by Antoine101** [mlflow/mlflow#11531] [Q&A]
Hi everyone,

I recently discovered it is possible to log system metrics with MLFlow. I tried to implement it both in my machine learning script and my lightning deep learning script but none of my attemps worked.
I followed instructions on this page: https://mlflow.org/docs/latest/system-metrics/index.html starting by installing the required packages.
(I have mlflow 2.11.3 - I am running on Windows 11)

I tried creating the environment variable by hand and also by adding this line of code: os.environ["MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING"] = "true"
Nothing worked...

At the execution of my machine learning script, I get prompted the following:
```
2024/03/26 10:15:38 INFO mlflow.system_metrics.system_metrics_monitor: Started monitoring system metrics.
C:\Users\APU\.conda\envs\alarm_ml\lib\importlib\__init__.py:126: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
  return _bootstrap._gcd_import(name[level:], package, level)
2024/03/26 10:15:42 INFO mlflow.system_metrics.system_metrics_monitor: Stopping system metrics monitoring...
2024/03/26 10:15:42 INFO mlflow.system_metrics.system_metrics_monitor: Successfully terminated system metrics monitoring!
```

But I have nothing in the system metrics tab:
![image](https://github.com/mlflow/mlflow/assets/48209504/000a29be-4b76-47d7-b893-d648475fca46)

Any idea?

Cheers

Antoine

**Comment by XnibyH** [comment]
Hello,
By default, the metrics are logged at 10-second intervals. I see your script ran for only 4 seconds, so no metrics were logged.

**Reply by theBeginner86** [reply]
Thanks!

This was helpful 👍

**Comment by SahilKumar75** [comment]
The existing comment nails the most common cause  if your run finishes in under 10 seconds, the default sampling interval never fires and nothing gets logged. But there are a few other things that can silently cause this on Windows, so worth covering all of them.

**1. Run duration shorter than the sampling interval**

The default `system_metrics_sampling_interval` is 10 seconds. If your script finishes in 4 seconds, the monitor starts, never gets a chance to collect a sample, and shuts down cleanly — which is exactly what your log shows:

Started monitoring system metrics.
Stopping system metrics monitoring...
Successfully terminated system metrics monitoring!

Fix: either lower the interval or make sure your run is long enough to get at least one sample.

```python
import mlflow

mlflow.set_system_metrics_sampling_interval(1)  # collect every 1 second
```

Or pass it directly to `start_run`:

```python
with mlflow.start_run(log_system_metrics=True) as run:
    mlflow.set_system_metrics_sampling_interval(1)
    # your training code
```

**2. Missing psutil (required dependency)**

System metrics on non-GPU machines rely on `psutil`. It's not always pulled in automatically. Install it if you haven't:

pip install psutil

For GPU metrics you also need `pynvml`:

pip install pynvml

If these aren't installed, MLflow silently skips those collectors.

**3. The env variable needs to be set before MLflow is imported**

This catches a lot of people. If you do:

```python
import mlflow
import os
os.environ["MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING"] = "true"
```

It won't work because MLflow already initialized without it. The env var needs to come first, or use the programmatic API instead which works at any point:

```python
import mlflow
mlflow.enable_system_metrics_logging()  # call this before start_run
```

**4. For PyTorch Lightning specifically**

If you're using the `MLFlowLogger`, system metrics logging isn't automatically inherited from the env var in all versions. Pass `log_system_metrics=True` explicitly:

```python
from lightning.pytorch.loggers import MLFlowLogger

logger = MLFlowLogger(
    experiment_name="my_experiment",
    log_system_metrics=True
)
```

**Quick checklist**

- `psutil` installed: `pip show psutil`
- Run lasts longer than the sampling interval (default 10s)
- `mlflow.enable_system_metrics_logging()` called before `start_run`, or `log_system_metrics=True` passed directly to `start_run`
- If using env var, set it before importing mlflow

The fact that you see "Started monitoring" and "Successfully terminated" in the logs means the feature is enabled and working it's almost certainly just the run duration issue combined with the 10-second default interval.
